homeassistant/components/aemet/__init__.py

```diff
@@ -0,0 +1,101 @@
+"""The AEMET OpenData component."""
+import asyncio
+import logging
+
+from aemet_opendata.interface import AEMET
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_API_KEY,
+    CONF_LATITUDE,
+    CONF_LONGITUDE,
+    CONF_MODE,
+    CONF_NAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import (
+    COMPONENTS,
+    DOMAIN,
+    ENTRY_NAME,
+    ENTRY_WEATHER_COORDINATOR,
+    UPDATE_LISTENER,
+)
+from .weather_update_coordinator import WeatherUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up the AEMET OpenData component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Set up AEMET OpenData as config entry."""
+    name = config_entry.data[CONF_NAME]
+    api_key = config_entry.data[CONF_API_KEY]
+    latitude = config_entry.data.get(CONF_LATITUDE, hass.config.latitude)
+    longitude = config_entry.data.get(CONF_LONGITUDE, hass.config.longitude)
+    forecast_mode = _get_config_value(config_entry, CONF_MODE)
+
+    aemet = AEMET(api_key)
+    weather_coordinator = WeatherUpdateCoordinator(
+        aemet, latitude, longitude, forecast_mode, hass
+    )
+
+    await weather_coordinator.async_refresh()
+
+    if not weather_coordinator.last_update_success:
+        raise ConfigEntryNotReady
```

Comment:
Is this needed? is it not possible to let the entry start and just let them end up being unavailable based on coordinator being unavailable?

Reply:
This is actually taken from OpenWeatherMap (https://github.com/home-assistant/core/blob/6b340415b2b6b3585d835f6e17aec7914fa29cd2/homeassistant/components/openweathermap/__init__.py#L59), but I will look into it.

Reply:
@elupus This should be solved with https://github.com/home-assistant/core/pull/45074/commits/5cce58b1460e5b1b9826914f0c59e92d140d0127

Reply:
Sort of yes. But now you will keep on logging an error on each failed request. Filling the log (also you can remove some checks for the return value of town == None).

You should either catch your new exception and rethrow a UpdateFailed exception. Or base the exception on UpdateFailed. That way the update coordinator will only log once on failure. See: https://github.com/home-assistant/core/blob/6f4cb18fa88d357eb873101b25159c989c0f9125/homeassistant/helpers/update_coordinator.py#L172

Reply:
@elupus done in https://github.com/home-assistant/core/pull/45074/commits/c96ed23bd03228ca5cb5176ce84d33e70a176442